cmd/dashboard/main.go

```diff
@@ -0,0 +1,275 @@
+		balance, nErr := d.l1Source.BalanceAt(d.ctx, d.l1Contract, new(big.Int).SetUint64(l.BlockNumber))
+		if nErr != nil {
+			return nil, start, fmt.Errorf("BalanceAt fail, block: %d, error: %s", l.BlockNumber, err.Error())
+		}
+		balanceBefore, oErr := d.l1Source.BalanceAt(d.ctx, d.l1Contract, new(big.Int).SetUint64(l.BlockNumber-1))
+		if oErr != nil {
+			return nil, start, fmt.Errorf("BalanceAt fail, block: %d, error: %s", l.BlockNumber-1, err.Error())
+		}
+		reward := new(big.Int).Sub(balanceBefore, balance)
+
+		events = append(events, &miningEvent{
+			ShardId:      new(big.Int).SetBytes(l.Topics[1].Bytes()).Uint64(),
+			Difficulty:   new(big.Int).SetBytes(l.Topics[2].Bytes()),
+			BlockMined:   new(big.Int).SetBytes(l.Topics[3].Bytes()),
+			LastMineTime: new(big.Int).SetBytes(l.Data[:32]).Uint64(),
+			// TODO: update when new version contract deployed, use the miner in the new MinedBlock event
+			Miner:  common.BytesToAddress(tx.Data()[80:100]),
+			Reward: reward.Uint64() / 10000000000,
+			GasFee: tx.Gas() * tx.GasPrice().Uint64() / 10000000000,
+		})
+	}
+	return events, end + 1, nil
+}
+
+func (d *dashboard) GetTransactionByHash(hash common.Hash) (*types.Transaction, error) {
+	tx, _, err := d.l1Source.TransactionByHash(context.Background(), hash)
+	if err != nil {
+		return nil, err
+	}
+	if len(tx.Data()) < 100 {
+		return nil, fmt.Errorf("invalid data len for tx %d", len(tx.Data()))
+	}
+	return tx, nil
+}
+
+func readSlotFromContract(ctx context.Context, client *ethclient.Client, l1Contract common.Address, fieldName string) ([]byte, error) {
+	h := crypto.Keccak256Hash([]byte(fieldName + "()"))
+	msg := ethereum.CallMsg{
+		To:   &l1Contract,
+		Data: h[0:4],
+	}
+	bs, err := client.CallContract(ctx, msg, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get %s from contract: %v", fieldName, err)
+	}
+	return bs, nil
+}
+
+func readUintFromContract(ctx context.Context, client *ethclient.Client, l1Contract common.Address, fieldName string) (uint64, error) {
+	bs, err := readSlotFromContract(ctx, client, l1Contract, fieldName)
+	if err != nil {
+		return 0, err
+	}
+	value := new(big.Int).SetBytes(bs).Uint64()
+	log.Info("Read uint from contract", "field", fieldName, "value", value)
+	return value, nil
+}
+
+func main() {
+	// Parse the flags and set up the logger to print everything requested
+	flag.Parse()
+	log.Root().SetHandler(log.LvlFilterHandler(log.Lvl(*logFlag), log.StreamHandler(os.Stderr, log.TerminalFormat(true))))
+
+	if *portFlag < 0 || *portFlag > math.MaxUint16 {
+		log.Crit("Invalid port")
+	}
+	d, err := newDashboard(*rpcURLFlag, common.HexToAddress(*l1ContractFlag))
+	if err != nil {
+		log.Crit("New dashboard fail", "err", err)
+	}
+
+	l1LatestBlockSub := eth.PollBlockChanges(d.ctx, d.logger, d.l1Source, d.RefreshBlobsMetrics, ethRPC.LatestBlockNumber,
+		10*time.Second, time.Second*10)
+	defer l1LatestBlockSub.Unsubscribe()
+
+	go d.FilterEventAndReport()
```

Comment:
I am not sure why we have two same interval timer here (one for RefreshBlobsMetrics, one for FilterEventAndReport ), can we combine into one?